twister unittest failing on python 3.12

<!--
**Notes**
Github Discussions (https://github.com/zephyrproject-rtos/zephyr/discussions)
are available to first verify that the issue is a genuine Zephyr bug and not a
consequence of Zephyr services misuse.

This issue list is only for bugs in the main Zephyr code base
(https://github.com/zephyrproject-rtos/zephyr/). If the bug is for a project
fork (such as NCS) specific feature, please open an issue in the fork project
instead.
-->

**Describe the bug**
<!--
A clear and concise description of what the bug is.

Please also mention any information which could help others to understand
the problem you're facing:
 - What target platform are you using?
 - What have you tried to diagnose or workaround this issue?
 - Is this a regression? If yes, have you been able to "git bisect" it to a
   specific commit?
 - ...
-->

see https://github.com/zephyrproject-rtos/zephyr/actions/runs/10309958475/job/28540679447?pr=74372, 

```
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED scripts/tests/twister/test_scl.py::test_yaml_imports[C YAML] - ModuleNotFoundError: spec not found for the module 'scl'
FAILED scripts/tests/twister/test_scl.py::test_yaml_imports[non-C YAML] - ModuleNotFoundError: spec not found for the module 'scl'
FAILED scripts/tests/twister/test_scl.py::test_pykwalify_import[pykwalify OK] - ModuleNotFoundError: spec not found for the module 'scl'
FAILED scripts/tests/twister/test_scl.py::test_pykwalify_import[no pykwalify] - ModuleNotFoundError: spec not found for the module 'scl'
FAILED scripts/tests/twister/test_testplan.py::test_required_snippets_short[app] - snippets.SnippetsError: error: invalid snippet.yml file: snippets/nordic-ppr/snippet.yml
  Schema validation failed:
   - Value '<Mock name='mock().get_single_data()' id='140417042117312'>' is not a dict. Value path: ''.
FAILED scripts/tests/twister/test_testplan.py::test_required_snippets_short[global] - snippets.SnippetsError: error: invalid snippet.yml file: snippets/nordic-ppr/snippet.yml
  Schema validation failed:
   - Value '<Mock name='mock().get_single_data()' id='140417042117312'>' is not a dict. Value path: ''.
FAILED scripts/tests/twister/test_testplan.py::test_required_snippets_short[multiple] - snippets.SnippetsError: error: invalid snippet.yml file: snippets/nordic-ppr/snippet.yml
  Schema validation failed:
   - Value '<Mock name='mock().get_single_data()' id='140417042117312'>' is not a dict. Value path: ''.
================= 7 failed, 704 passed, 468 warnings in 10.86s =================
Error: Process completed with exit code 1.
```
**To Reproduce**
<!--
Steps to reproduce the behavior:
1. mkdir build; cd build
2. cmake -DBOARD=board\_xyz
3. make
4. See error
-->

submit a change related to twister

**Expected behavior**
<!--
A clear and concise description of what you expected to happen.
-->

tests should pass.

**Impact**
<!--
What impact does this issue have on your progress (e.g., annoyance, showstopper)
-->

All twister related PRs are currently being blocked by this check.

--- a/scripts/pylib/twister/twisterlib/environment.py
+++ b/scripts/pylib/twister/twisterlib/environment.py
@@ -5,17 +5,19 @@
 # Copyright 2022 NXP
 # SPDX-License-Identifier: Apache-2.0
 
-import os
-import pkg_resources
-import sys
-from pathlib import Path
+import argparse
 import json
 import logging
-import subprocess
-import shutil
+import os
 import re
-import argparse
+import shutil
+import subprocess
+import sys
 from datetime import datetime, timezone
+from importlib import metadata
+from pathlib import Path
+from typing import Generator
+
 from twisterlib.coverage import supported_coverage_formats
 
 logger = logging.getLogger('twister')
@@ -40,12 +42,21 @@ import zephyr_module
 # Note "normalization" is different from canonicalization, see os.path.
 canonical_zephyr_base = os.path.realpath(ZEPHYR_BASE)
 
-installed_packages = [pkg.project_name for pkg in pkg_resources.working_set]  # pylint: disable=not-an-iterable
+
+def _get_installed_packages() -> Generator[str, None, None]:
+    """Return list of installed python packages."""
+    for dist in metadata.distributions():
+        yield dist.metadata['Name']
+
+
+installed_packages: list[str] = list(_get_installed_packages())
 PYTEST_PLUGIN_INSTALLED = 'pytest-twister-harness' in installed_packages
+
 
 def norm_path(astring):
     newstring = os.path.normpath(astring).replace(os.sep, '/')
     return newstring
+
 
 def add_parse_arguments(parser = None):
     if parser is None:

--- a/scripts/pylib/twister/twisterlib/harness.py
+++ b/scripts/pylib/twister/twisterlib/harness.py
@@ -668,6 +668,7 @@ class Gtest(Harness):
 
 
 class Test(Harness):
+    __test__ = False  # for pytest to skip this class when collects tests
     RUN_PASSED = "PROJECT EXECUTION SUCCESSFUL"
     RUN_FAILED = "PROJECT EXECUTION FAILED"
     test_suite_start_pattern = r"Running TESTSUITE (?P<suite_name>.*)"

--- a/scripts/pylib/twister/twisterlib/testplan.py
+++ b/scripts/pylib/twister/twisterlib/testplan.py
@@ -77,7 +77,9 @@ class TestLevel:
     levels = []
     scenarios = []
 
+
 class TestPlan:
+    __test__ = False  # for pytest to skip this class when collects tests
     config_re = re.compile('(CONFIG_[A-Za-z0-9_]+)[=]\"?([^\"]*)\"?$')
     dt_re = re.compile('([A-Za-z0-9_]+)[=]\"?([^\"]*)\"?$')
 

